swagger-ui endpoint not accessible in AWS Lambda

**Describe the bug**
Making a http call to /swagger-ui/ , the app fails to serve the swagger-ui files. Local Dev mode works fine. 

**Expected behavior**
Opening /swagger-ui/ should load swagger-ui/index.html page

**Actual behavior**
ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-2) HTTP Request to /swagger-ui/index.html failed, error id: bea62c51-a11f-4cb4-8aba-86217c872b1d-1: io.vertx.core.file.FileSystemException: java.nio.file.NoSuchFileException: META-INF/swagger-ui-files/index.html
	at io.vertx.core.file.impl.FileSystemImpl$7.perform(FileSystemImpl.java:625)
	at io.vertx.core.file.impl.FileSystemImpl$7.perform(FileSystemImpl.java:613)
	at io.vertx.core.file.impl.FileSystemImpl.propsBlocking(FileSystemImpl.java:162)
	at io.vertx.ext.web.handler.impl.StaticHandlerImpl.lambda$getFileProps$4(StaticHandlerImpl.java:295)
	at io.vertx.ext.web.handler.impl.StaticHandlerImpl.wrapInTCCLSwitch(StaticHandlerImpl.java:266)
	at io.vertx.ext.web.handler.impl.StaticHandlerImpl.getFileProps(StaticHandlerImpl.java:295)
	at io.vertx.ext.web.handler.impl.StaticHandlerImpl.lambda$sendStatic$1(StaticHandlerImpl.java:208)
	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:330)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:369)
	at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.nio.file.NoSuchFileException: META-INF/swagger-ui-files/index.html
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
	at sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:144)
	at sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
	at java.nio.file.Files.readAttributes(Files.java:1737)
	at io.vertx.core.file.impl.FileSystemImpl$7.perform(FileSystemImpl.java:619)


**To Reproduce**
Steps to reproduce the behavior:
1.  Create a sample Quarkus project with Quarkus 1.3.0.final 
2.  Maven Dependencies - quarkus-resteasy, resteasy-jackson, quarkus-smallrye-openapi and quarkus-amazon-lambda-http
3. SAM template - Lambda to API gateway proxy integration 
4. Set quarkus.swagger-ui.always-include=true in application.properities
5. Deploy the app to lambda 
6.) Try to accessing <host>/swagger-ui/ endpoint 

**Configuration**
```properties
# Add your application.properties here, if applicable.
quarkus.swagger-ui.always-include=true
```

**Screenshots**
(If applicable, add screenshots to help explain your problem.)

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
 - Output of `java -version`:  java8
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 1.3.0.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): maven 3.6.2+

**Additional context**
(Add any other context about the problem here.)

--- a/extensions/amazon-lambda-http/runtime/src/main/java/io/quarkus/amazon/lambda/http/LambdaHttpHandler.java
+++ b/extensions/amazon-lambda-http/runtime/src/main/java/io/quarkus/amazon/lambda/http/LambdaHttpHandler.java
@@ -3,11 +3,15 @@ package io.quarkus.amazon.lambda.http;
 import java.io.ByteArrayOutputStream;
 import java.net.InetSocketAddress;
 import java.net.URLEncoder;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
 import java.nio.charset.StandardCharsets;
 import java.util.Base64;
 import java.util.List;
 import java.util.Map;
-import java.util.concurrent.TimeUnit;
+import java.util.concurrent.CompletableFuture;
+
+import org.jboss.logging.Logger;
 
 import com.amazonaws.serverless.proxy.internal.LambdaContainerHandler;
 import com.amazonaws.services.lambda.runtime.Context;
@@ -15,6 +19,7 @@ import com.amazonaws.services.lambda.runtime.RequestHandler;
 
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.Unpooled;
+import io.netty.channel.FileRegion;
 import io.netty.handler.codec.http.DefaultHttpRequest;
 import io.netty.handler.codec.http.DefaultLastHttpContent;
 import io.netty.handler.codec.http.HttpContent;
@@ -28,10 +33,12 @@ import io.quarkus.amazon.lambda.http.model.AwsProxyRequest;
 import io.quarkus.amazon.lambda.http.model.AwsProxyResponse;
 import io.quarkus.amazon.lambda.http.model.Headers;
 import io.quarkus.netty.runtime.virtual.VirtualClientConnection;
+import io.quarkus.netty.runtime.virtual.VirtualResponseHandler;
 import io.quarkus.vertx.http.runtime.VertxHttpRecorder;
 
 @SuppressWarnings("unused")
 public class LambdaHttpHandler implements RequestHandler<AwsProxyRequest, AwsProxyResponse> {
+    private static final Logger log = Logger.getLogger("quarkus.amazon.lambda.http");
 
     private static Headers errorHeaders = new Headers();
     static {
@@ -46,19 +53,99 @@ public class LambdaHttpHandler implements RequestHandler<AwsProxyRequest, AwsPro
             }
         }
 
-        VirtualClientConnection connection = VirtualClientConnection.connect(VertxHttpRecorder.VIRTUAL_HTTP, clientAddress);
         try {
-            return nettyDispatch(connection, request);
+            return nettyDispatch(clientAddress, request);
         } catch (Exception e) {
+            log.error("Request Failure", e);
             return new AwsProxyResponse(500, errorHeaders, "{ \"message\": \"Internal Server Error\" }");
-        } finally {
-            connection.close();
         }
 
     }
 
-    private AwsProxyResponse nettyDispatch(VirtualClientConnection connection, AwsProxyRequest request) throws Exception {
+    private class NettyResponseHandler implements VirtualResponseHandler {
+        AwsProxyResponse responseBuilder = new AwsProxyResponse();
+        ByteArrayOutputStream baos;
+        WritableByteChannel byteChannel;
+        final AwsProxyRequest request;
+        CompletableFuture<AwsProxyResponse> future = new CompletableFuture<>();
+
+        public NettyResponseHandler(AwsProxyRequest request) {
+            this.request = request;
+        }
+
+        public CompletableFuture<AwsProxyResponse> getFuture() {
+            return future;
+        }
+
+        @Override
+        public void handleMessage(Object msg) {
+            try {
+                //log.info("Got message: " + msg.getClass().getName());
+
+                if (msg instanceof HttpResponse) {
+                    HttpResponse res = (HttpResponse) msg;
+                    responseBuilder.setStatusCode(res.status().code());
+
+                    if (request.getRequestSource() == AwsProxyRequest.RequestSource.ALB) {
+                        responseBuilder.setStatusDescription(res.status().reasonPhrase());
+                    }
+                    responseBuilder.setMultiValueHeaders(new Headers());
+                    for (String name : res.headers().names()) {
+                        for (String v : res.headers().getAll(name)) {
+                            responseBuilder.getMultiValueHeaders().add(name, v);
+                        }
+                    }
+                }
+                if (msg instanceof HttpContent) {
+                    HttpContent content = (HttpContent) msg;
+                    int readable = content.content().readableBytes();
+                    if (baos == null && readable > 0) {
+                        baos = createByteStream();
+                    }
+                    for (int i = 0; i < readable; i++) {
+                        baos.write(content.content().readByte());
+                    }
+                }
+                if (msg instanceof FileRegion) {
+                    FileRegion file = (FileRegion) msg;
+                    if (file.count() > 0 && file.transferred() < file.count()) {
+                        if (baos == null)
+                            baos = createByteStream();
+                        if (byteChannel == null)
+                            byteChannel = Channels.newChannel(baos);
+                        file.transferTo(byteChannel, file.transferred());
+                    }
+                }
+                if (msg instanceof LastHttpContent) {
+                    if (baos != null) {
+                        if (isBinary(responseBuilder.getMultiValueHeaders().getFirst("Content-Type"))) {
+                            responseBuilder.setBase64Encoded(true);
+                            responseBuilder.setBody(Base64.getMimeEncoder().encodeToString(baos.toByteArray()));
+                        } else {
+                            responseBuilder.setBody(new String(baos.toByteArray(), "UTF-8"));
+                        }
+                    }
+                    future.complete(responseBuilder);
+                }
+            } catch (Throwable ex) {
+                future.completeExceptionally(ex);
+            } finally {
+                if (msg != null) {
+                    ReferenceCountUtil.release(msg);
+                }
+            }
+        }
+
+        @Override
+        public void close() {
+            if (!future.isDone())
+                future.completeExceptionally(new RuntimeException("Connection closed"));
+        }
+    }
+
+    private AwsProxyResponse nettyDispatch(InetSocketAddress clientAddress, AwsProxyRequest request) throws Exception {
         String path = request.getPath();
+        //log.info("---- Got lambda request: " + path);
         if (request.getMultiValueQueryStringParameters() != null && !request.getMultiValueQueryStringParameters().isEmpty()) {
             StringBuilder sb = new StringBuilder(path);
             sb.append("?");
@@ -104,67 +191,23 @@ public class LambdaHttpHandler implements RequestHandler<AwsProxyRequest, AwsPro
                 requestContent = new DefaultLastHttpContent(body);
             }
         }
+        NettyResponseHandler handler = new NettyResponseHandler(request);
+        VirtualClientConnection connection = VirtualClientConnection.connect(handler, VertxHttpRecorder.VIRTUAL_HTTP,
+                clientAddress);
 
         connection.sendMessage(nettyRequest);
         connection.sendMessage(requestContent);
-        AwsProxyResponse responseBuilder = new AwsProxyResponse();
-        ByteArrayOutputStream baos = null;
         try {
-            for (;;) {
-                // todo should we timeout? have a timeout config?
-                //log.info("waiting for message");
-                Object msg = connection.queue().poll(100, TimeUnit.MILLISECONDS);
-                try {
-                    if (msg == null)
-                        continue;
-                    //log.info("Got message: " + msg.getClass().getName());
-
-                    if (msg instanceof HttpResponse) {
-                        HttpResponse res = (HttpResponse) msg;
-                        responseBuilder.setStatusCode(res.status().code());
-
-                        if (request.getRequestSource() == AwsProxyRequest.RequestSource.ALB) {
-                            responseBuilder.setStatusDescription(res.status().reasonPhrase());
-                        }
-                        responseBuilder.setMultiValueHeaders(new Headers());
-                        for (String name : res.headers().names()) {
-                            for (String v : res.headers().getAll(name)) {
-                                responseBuilder.getMultiValueHeaders().add(name, v);
-                            }
-                        }
-                    }
-                    if (msg instanceof HttpContent) {
-                        HttpContent content = (HttpContent) msg;
-                        int readable = content.content().readableBytes();
-                        if (baos == null && readable > 0) {
-                            // todo what is right size?
-                            baos = new ByteArrayOutputStream(500);
-                        }
-                        for (int i = 0; i < readable; i++) {
-                            baos.write(content.content().readByte());
-                        }
-                    }
-                    if (msg instanceof LastHttpContent) {
-                        if (baos != null) {
-                            if (isBinary(responseBuilder.getMultiValueHeaders().getFirst("Content-Type"))) {
-                                responseBuilder.setBase64Encoded(true);
-                                responseBuilder.setBody(Base64.getMimeEncoder().encodeToString(baos.toByteArray()));
-                            } else {
-                                responseBuilder.setBody(new String(baos.toByteArray(), "UTF-8"));
-                            }
-                        }
-                        return responseBuilder;
-                    }
-                } finally {
-                    if (msg != null)
-                        ReferenceCountUtil.release(msg);
-                }
-            }
+            return handler.getFuture().get();
         } finally {
-            if (baos != null) {
-                baos.close();
-            }
+            connection.close();
         }
+    }
+
+    private ByteArrayOutputStream createByteStream() {
+        ByteArrayOutputStream baos;// todo what is right size?
+        baos = new ByteArrayOutputStream(1000);
+        return baos;
     }
 
     private boolean isBinary(String contentType) {

--- a/extensions/amazon-lambda/common-runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AbstractLambdaPollLoop.java
+++ b/extensions/amazon-lambda/common-runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AbstractLambdaPollLoop.java
@@ -3,7 +3,9 @@ package io.quarkus.amazon.lambda.runtime;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.net.ConnectException;
 import java.net.HttpURLConnection;
+import java.net.SocketException;
 import java.net.URL;
 import java.util.concurrent.atomic.AtomicBoolean;
 
@@ -45,7 +47,15 @@ public abstract class AbstractLambdaPollLoop {
                     URL requestUrl = AmazonLambdaApi.invocationNext();
                     while (running.get()) {
 
-                        HttpURLConnection requestConnection = (HttpURLConnection) requestUrl.openConnection();
+                        HttpURLConnection requestConnection = null;
+                        try {
+                            requestConnection = (HttpURLConnection) requestUrl.openConnection();
+                        } catch (IOException e) {
+                            if (abortGracefully(e)) {
+                                return;
+                            }
+                            throw e;
+                        }
                         try {
                             String requestId = requestConnection.getHeaderField(AmazonLambdaApi.LAMBDA_RUNTIME_AWS_REQUEST_ID);
                             try {
@@ -70,6 +80,9 @@ public abstract class AbstractLambdaPollLoop {
                                     }
                                 }
                             } catch (Exception e) {
+                                if (abortGracefully(e)) {
+                                    return;
+                                }
                                 log.error("Failed to run lambda", e);
 
                                 postError(AmazonLambdaApi.invocationError(requestId),
@@ -78,7 +91,8 @@ public abstract class AbstractLambdaPollLoop {
                             }
 
                         } catch (Exception e) {
-                            log.error("Error running lambda", e);
+                            if (!abortGracefully(e))
+                                log.error("Error running lambda", e);
                             Application app = Application.currentApplication();
                             if (app != null) {
                                 app.stop();
@@ -172,6 +186,16 @@ public abstract class AbstractLambdaPollLoop {
         responseConnection.setDoOutput(true);
         responseConnection.setRequestMethod("POST");
         return responseConnection;
+    }
+
+    boolean abortGracefully(Exception ex) {
+        // if we are running in test mode, then don't output stack trace for socket errors
+
+        boolean graceful = (ex instanceof SocketException || ex instanceof ConnectException)
+                && System.getProperty(AmazonLambdaApi.QUARKUS_INTERNAL_AWS_LAMBDA_TEST_API) != null;
+        if (graceful)
+            log.warn("Aborting lambda poll loop");
+        return graceful;
     }
 
 }

--- a/extensions/azure-functions-http/runtime/src/main/java/io/quarkus/azure/functions/resteasy/runtime/BaseFunction.java
+++ b/extensions/azure-functions-http/runtime/src/main/java/io/quarkus/azure/functions/resteasy/runtime/BaseFunction.java
@@ -3,9 +3,11 @@ package io.quarkus.azure.functions.resteasy.runtime;
 import java.io.ByteArrayOutputStream;
 import java.io.PrintWriter;
 import java.io.StringWriter;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
 import java.util.Map;
 import java.util.Optional;
-import java.util.concurrent.TimeUnit;
+import java.util.concurrent.CompletableFuture;
 
 import org.jboss.logging.Logger;
 
@@ -15,6 +17,7 @@ import com.microsoft.azure.functions.HttpStatus;
 
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.Unpooled;
+import io.netty.channel.FileRegion;
 import io.netty.handler.codec.http.DefaultHttpRequest;
 import io.netty.handler.codec.http.DefaultLastHttpContent;
 import io.netty.handler.codec.http.HttpContent;
@@ -24,6 +27,7 @@ import io.netty.handler.codec.http.HttpVersion;
 import io.netty.handler.codec.http.LastHttpContent;
 import io.netty.util.ReferenceCountUtil;
 import io.quarkus.netty.runtime.virtual.VirtualClientConnection;
+import io.quarkus.netty.runtime.virtual.VirtualResponseHandler;
 import io.quarkus.runtime.Application;
 import io.quarkus.vertx.http.runtime.VertxHttpRecorder;
 
@@ -56,20 +60,16 @@ public class BaseFunction {
     }
 
     protected HttpResponseMessage dispatch(HttpRequestMessage<Optional<byte[]>> request) {
-        VirtualClientConnection connection = VirtualClientConnection.connect(VertxHttpRecorder.VIRTUAL_HTTP);
         try {
-            return nettyDispatch(connection, request);
+            return nettyDispatch(request);
         } catch (Exception e) {
             e.printStackTrace();
             return request
                     .createResponseBuilder(HttpStatus.valueOf(500)).build();
-        } finally {
-            connection.close();
         }
     }
 
-    protected HttpResponseMessage nettyDispatch(VirtualClientConnection connection,
-            HttpRequestMessage<Optional<byte[]>> request)
+    protected HttpResponseMessage nettyDispatch(HttpRequestMessage<Optional<byte[]>> request)
             throws Exception {
         String path = request.getUri().getRawPath();
         String query = request.getUri().getRawQuery();
@@ -92,51 +92,83 @@ public class BaseFunction {
             requestContent = new DefaultLastHttpContent(body);
         }
 
+        ResponseHandler handler = new ResponseHandler(request);
+        VirtualClientConnection connection = VirtualClientConnection.connect(handler, VertxHttpRecorder.VIRTUAL_HTTP);
+
         connection.sendMessage(nettyRequest);
         connection.sendMessage(requestContent);
-        HttpResponseMessage.Builder responseBuilder = null;
-        ByteArrayOutputStream baos = null;
         try {
-            for (;;) {
-                // todo should we timeout? have a timeout config?
-                //log.info("waiting for message");
-                Object msg = connection.queue().poll(100, TimeUnit.MILLISECONDS);
-                try {
-                    if (msg == null)
-                        continue;
-                    //log.info("Got message: " + msg.getClass().getName());
-
-                    if (msg instanceof HttpResponse) {
-                        HttpResponse res = (HttpResponse) msg;
-                        responseBuilder = request.createResponseBuilder(HttpStatus.valueOf(res.status().code()));
-                        for (Map.Entry<String, String> entry : res.headers()) {
-                            responseBuilder.header(entry.getKey(), entry.getValue());
-                        }
-                    }
-                    if (msg instanceof HttpContent) {
-                        HttpContent content = (HttpContent) msg;
-                        if (baos == null) {
-                            // todo what is right size?
-                            baos = new ByteArrayOutputStream(500);
-                        }
-                        int readable = content.content().readableBytes();
-                        for (int i = 0; i < readable; i++) {
-                            baos.write(content.content().readByte());
-                        }
-                    }
-                    if (msg instanceof LastHttpContent) {
-                        responseBuilder.body(baos.toByteArray());
-                        return responseBuilder.build();
-                    }
-                } finally {
-                    if (msg != null)
-                        ReferenceCountUtil.release(msg);
-                }
-            }
+            return handler.future.get();
         } finally {
-            if (baos != null) {
-                baos.close();
+            connection.close();
+        }
+    }
+
+    private ByteArrayOutputStream createByteStream() {
+        ByteArrayOutputStream baos;
+        baos = new ByteArrayOutputStream(500);
+        return baos;
+    }
+
+    private class ResponseHandler implements VirtualResponseHandler {
+        HttpResponseMessage.Builder responseBuilder;
+        ByteArrayOutputStream baos;
+        WritableByteChannel byteChannel;
+        CompletableFuture<HttpResponseMessage> future = new CompletableFuture<>();
+        final HttpRequestMessage<Optional<byte[]>> request;
+
+        public ResponseHandler(HttpRequestMessage<Optional<byte[]>> request) {
+            this.request = request;
+        }
+
+        @Override
+        public void handleMessage(Object msg) {
+            try {
+                //log.info("Got message: " + msg.getClass().getName());
+
+                if (msg instanceof HttpResponse) {
+                    HttpResponse res = (HttpResponse) msg;
+                    responseBuilder = request.createResponseBuilder(HttpStatus.valueOf(res.status().code()));
+                    for (Map.Entry<String, String> entry : res.headers()) {
+                        responseBuilder.header(entry.getKey(), entry.getValue());
+                    }
+                }
+                if (msg instanceof HttpContent) {
+                    HttpContent content = (HttpContent) msg;
+                    if (baos == null) {
+                        // todo what is right size?
+                        baos = createByteStream();
+                    }
+                    int readable = content.content().readableBytes();
+                    for (int i = 0; i < readable; i++) {
+                        baos.write(content.content().readByte());
+                    }
+                }
+                if (msg instanceof FileRegion) {
+                    FileRegion file = (FileRegion) msg;
+                    if (file.count() > 0 && file.transferred() < file.count()) {
+                        if (baos == null)
+                            baos = createByteStream();
+                        if (byteChannel == null)
+                            byteChannel = Channels.newChannel(baos);
+                        file.transferTo(byteChannel, file.transferred());
+                    }
+                }
+                if (msg instanceof LastHttpContent) {
+                    responseBuilder.body(baos.toByteArray());
+                    future.complete(responseBuilder.build());
+                }
+            } catch (Throwable ex) {
+                future.completeExceptionally(ex);
+            } finally {
+                ReferenceCountUtil.release(msg);
             }
+        }
+
+        @Override
+        public void close() {
+            if (!future.isDone())
+                future.completeExceptionally(new RuntimeException("Connection closed"));
         }
     }
 }

--- a/extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/virtual/VirtualChannel.java
+++ b/extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/virtual/VirtualChannel.java
@@ -313,7 +313,8 @@ public class VirtualChannel extends AbstractChannel {
                     // It is possible the peer could have closed while we are writing, and in this case we should
                     // simulate real socket behavior and ensure the sendMessage operation is failed.
                     if (peer.isConnected()) {
-                        peer.queue().add(ReferenceCountUtil.retain(msg));
+                        ReferenceCountUtil.retain(msg);
+                        peer.handler.handleMessage(msg);
                         in.remove();
                     } else {
                         if (exception == null) {

--- a/extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/virtual/VirtualClientConnection.java
+++ b/extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/virtual/VirtualClientConnection.java
@@ -1,46 +1,34 @@
 package io.quarkus.netty.runtime.virtual;
 
 import java.net.SocketAddress;
-import java.util.concurrent.BlockingQueue;
-import java.util.concurrent.LinkedBlockingQueue;
 
 import io.netty.channel.Channel;
 import io.netty.util.concurrent.Future;
 import io.netty.util.internal.PlatformDependent;
 
 /**
- * A virtual client connection to an intra-JVM netty server channel.
- *
- * Clients can block and wait directly for posted messages from the server channel.
- *
+ * A virtual client connection to an intra-JVM request/response netty server channel.
  */
-public class VirtualClientConnection {
-    protected SocketAddress clientAddress;
-    protected BlockingQueue<Object> queue = new LinkedBlockingQueue<>();
+public class VirtualClientConnection<T> {
+    protected final SocketAddress clientAddress;
     protected boolean connected = true;
     protected VirtualChannel peer;
+    protected final VirtualResponseHandler handler;
 
-    VirtualClientConnection(SocketAddress clientAddress) {
+    public VirtualClientConnection(SocketAddress clientAddress, VirtualResponseHandler handler) {
         this.clientAddress = clientAddress;
+        this.handler = handler;
     }
 
     public SocketAddress clientAddress() {
         return clientAddress;
     }
 
-    /**
-     * Blocking queue for a client to obtain messages directly from server
-     *
-     * @return
-     */
-    public BlockingQueue<Object> queue() {
-        return queue;
-    }
-
     public void close() {
         // todo more cleanup?
         connected = false;
         peer.close();
+        handler.close();
     }
 
     public boolean isConnected() {
@@ -104,9 +92,8 @@ public class VirtualClientConnection {
      * @param remoteAddress
      * @return
      */
-    public static VirtualClientConnection connect(final VirtualAddress remoteAddress) {
-        return connect(remoteAddress, remoteAddress);
-
+    public static VirtualClientConnection connect(VirtualResponseHandler handler, VirtualAddress remoteAddress) {
+        return connect(handler, remoteAddress, remoteAddress);
     }
 
     /**
@@ -116,7 +103,8 @@ public class VirtualClientConnection {
      * @param clientAddress
      * @return
      */
-    public static VirtualClientConnection connect(VirtualAddress remoteAddress, SocketAddress clientAddress) {
+    public static VirtualClientConnection connect(VirtualResponseHandler handler, VirtualAddress remoteAddress,
+            SocketAddress clientAddress) {
         if (clientAddress == null)
             clientAddress = remoteAddress;
 
@@ -129,7 +117,7 @@ public class VirtualClientConnection {
         }
 
         VirtualServerChannel serverChannel = (VirtualServerChannel) boundChannel;
-        VirtualClientConnection conn = new VirtualClientConnection(clientAddress);
+        VirtualClientConnection conn = new VirtualClientConnection(clientAddress, handler);
         conn.peer = serverChannel.serve(conn);
         return conn;
     }

--- a/extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/virtual/VirtualResponseHandler.java
+++ b/extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/virtual/VirtualResponseHandler.java
@@ -1,0 +1,7 @@
+package io.quarkus.netty.runtime.virtual;
+
+public interface VirtualResponseHandler {
+    void handleMessage(Object msg);
+
+    void close();
+}

--- a/integration-tests/amazon-lambda-http/pom.xml
+++ b/integration-tests/amazon-lambda-http/pom.xml
@@ -24,6 +24,10 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-smallrye-openapi</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-undertow</artifactId>
         </dependency>
         <dependency>

--- a/integration-tests/amazon-lambda-http/src/main/resources/application.properties
+++ b/integration-tests/amazon-lambda-http/src/main/resources/application.properties
@@ -1,2 +1,3 @@
 quarkus.lambda.enable-polling-jvm-mode=true
 quarkus.http.virtual=true
+quarkus.swagger-ui.always-include=true

--- a/integration-tests/amazon-lambda-http/src/test/java/io/quarkus/it/amazon/lambda/AmazonLambdaSimpleTestCase.java
+++ b/integration-tests/amazon-lambda-http/src/test/java/io/quarkus/it/amazon/lambda/AmazonLambdaSimpleTestCase.java
@@ -24,6 +24,16 @@ public class AmazonLambdaSimpleTestCase {
         testGetText("/hello");
     }
 
+    @Test
+    public void testSwaggerUi() throws Exception {
+        // this tests the FileRegion support in the handler
+        AwsProxyRequest request = request("/swagger-ui/");
+        AwsProxyResponse out = LambdaClient.invoke(AwsProxyResponse.class, request);
+        Assertions.assertEquals(out.getStatusCode(), 200);
+        Assertions.assertTrue(body(out).contains("Swagger UI"));
+
+    }
+
     private String body(AwsProxyResponse response) {
         if (!response.isBase64Encoded())
             return response.getBody();
@@ -31,20 +41,23 @@ public class AmazonLambdaSimpleTestCase {
     }
 
     private void testGetText(String path) {
-        AwsProxyRequest request = new AwsProxyRequest();
-        request.setHttpMethod("GET");
-        request.setPath(path);
+        AwsProxyRequest request = request(path);
         AwsProxyResponse out = LambdaClient.invoke(AwsProxyResponse.class, request);
         Assertions.assertEquals(out.getStatusCode(), 200);
         Assertions.assertEquals(body(out), "hello");
         Assertions.assertTrue(out.getMultiValueHeaders().getFirst("Content-Type").startsWith("text/plain"));
     }
 
-    @Test
-    public void test404() throws Exception {
+    private AwsProxyRequest request(String path) {
         AwsProxyRequest request = new AwsProxyRequest();
         request.setHttpMethod("GET");
-        request.setPath("/nowhere");
+        request.setPath(path);
+        return request;
+    }
+
+    @Test
+    public void test404() throws Exception {
+        AwsProxyRequest request = request("/nowhere");
         AwsProxyResponse out = LambdaClient.invoke(AwsProxyResponse.class, request);
         Assertions.assertEquals(out.getStatusCode(), 404);
     }

--- a/integration-tests/virtual-http-resteasy/pom.xml
+++ b/integration-tests/virtual-http-resteasy/pom.xml
@@ -24,6 +24,10 @@
             <artifactId>quarkus-azure-functions-http</artifactId>
         </dependency>
         <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-smallrye-openapi</artifactId>
+        </dependency>
+        <dependency>
             <groupId>com.microsoft.azure.functions</groupId>
             <artifactId>azure-functions-java-library</artifactId>
             <scope>test</scope>

--- a/integration-tests/virtual-http-resteasy/src/main/resources/application.properties
+++ b/integration-tests/virtual-http-resteasy/src/main/resources/application.properties
@@ -1,1 +1,3 @@
 quarkus.http.virtual=true
+quarkus.swagger-ui.always-include=true
+

--- a/integration-tests/virtual-http-resteasy/src/test/java/io/quarkus/it/virtual/FunctionTest.java
+++ b/integration-tests/virtual-http-resteasy/src/test/java/io/quarkus/it/virtual/FunctionTest.java
@@ -27,20 +27,22 @@ import io.restassured.RestAssured;
 @QuarkusTest
 public class FunctionTest {
     @Test
-    public void testJaxrs() throws Exception {
-        String uri = "https://foo.com/hello";
-        testGET(uri);
-        testPOST(uri);
-    }
-
-    @Test
-    public void testNotFound() {
+    public void testSwagger() {
         final HttpRequestMessageMock req = new HttpRequestMessageMock();
-        req.setUri(URI.create("https://nowhere.com/badroute"));
+        req.setUri(URI.create("https://foo.com/swagger-ui/"));
         req.setHttpMethod(HttpMethod.GET);
 
         // Invoke
-        final HttpResponseMessage ret = new Function().run(req, new ExecutionContext() {
+        final HttpResponseMessage ret = new Function().run(req, createContext());
+
+        // Verify
+        Assertions.assertEquals(ret.getStatus(), HttpStatus.OK);
+        String body = new String((byte[]) ret.getBody(), StandardCharsets.UTF_8);
+        Assertions.assertTrue(body.contains("Swagger UI"));
+    }
+
+    private ExecutionContext createContext() {
+        return new ExecutionContext() {
             @Override
             public Logger getLogger() {
                 return null;
@@ -55,7 +57,24 @@ public class FunctionTest {
             public String getFunctionName() {
                 return null;
             }
-        });
+        };
+    }
+
+    @Test
+    public void testJaxrs() throws Exception {
+        String uri = "https://foo.com/hello";
+        testGET(uri);
+        testPOST(uri);
+    }
+
+    @Test
+    public void testNotFound() {
+        final HttpRequestMessageMock req = new HttpRequestMessageMock();
+        req.setUri(URI.create("https://nowhere.com/badroute"));
+        req.setHttpMethod(HttpMethod.GET);
+
+        // Invoke
+        final HttpResponseMessage ret = new Function().run(req, createContext());
 
         // Verify
         Assertions.assertEquals(ret.getStatus(), HttpStatus.NOT_FOUND);
@@ -79,22 +98,7 @@ public class FunctionTest {
         req.setHttpMethod(HttpMethod.GET);
 
         // Invoke
-        final HttpResponseMessage ret = new Function().run(req, new ExecutionContext() {
-            @Override
-            public Logger getLogger() {
-                return null;
-            }
-
-            @Override
-            public String getInvocationId() {
-                return null;
-            }
-
-            @Override
-            public String getFunctionName() {
-                return null;
-            }
-        });
+        final HttpResponseMessage ret = new Function().run(req, createContext());
 
         // Verify
         Assertions.assertEquals(ret.getStatus(), HttpStatus.OK);
@@ -112,22 +116,7 @@ public class FunctionTest {
         req.getHeaders().put("Content-Type", "text/plain");
 
         // Invoke
-        final HttpResponseMessage ret = new Function().run(req, new ExecutionContext() {
-            @Override
-            public Logger getLogger() {
-                return null;
-            }
-
-            @Override
-            public String getInvocationId() {
-                return null;
-            }
-
-            @Override
-            public String getFunctionName() {
-                return null;
-            }
-        });
+        final HttpResponseMessage ret = new Function().run(req, createContext());
 
         // Verify
         Assertions.assertEquals(ret.getStatus(), HttpStatus.OK);

